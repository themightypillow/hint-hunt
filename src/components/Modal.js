import React from "react";
import { observer } from "mobx-react";
import Loader from "./Loader";

@observer
class Modal extends React.Component {
  render() {
    const store = this.props.store;
    return (
      <div className="modal">
        {
          store.modal.loading ?
            <div className="modal_text"><Loader /></div> :
            <div className="modal_text">
              {store.modal.text}
              {
                store.modal.button && 
                <button onClick={store.modal.hide}>{store.modal.buttonText}</button>
              }
            </div>
        }
      </div>
    );
  }
}

export default Modal;