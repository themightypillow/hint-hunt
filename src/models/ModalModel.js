import { configure, observable, action } from "mobx";

// state can only be changed inside @action methods
configure({ enforceActions: "observed" });

class ModalModel {

  @observable visible = false;

  @observable heading = "";

  @observable loading = false;

  // Date object once set
  @observable date = null;

  @observable buttonDisabled = false;

  @action hide = () => {
    this.visible = false;
  };

  @action hideLoading = () => {
    this.loading = false;
  };

  @action setHeading = (title) => {
    this.heading = `${this.date.toDateString()} - ${title}`;
  };

  // Only runs after the user finishes/wins a word search
  @action.bound showCalendar = () => {  
    this.visible = true;
    this.heading = "Choose a Date";
    this.loading = false;
    this.date = null;
    this.buttonDisabled = true;
    
    // 'reset' the animated win screen so that it can be animated again the future
    let animateScreen = document.querySelector(".hinthunt_animate");
    const promise = new Promise((resolve) => {
      setTimeout(() => {
        animateScreen.classList.remove("hinthunt_animate-on", "hinthunt_animate-visible");
        resolve();
      }, 500);
    });
    return promise;
  }
}

export default ModalModel;