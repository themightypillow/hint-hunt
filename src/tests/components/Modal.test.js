import React from "react";
import { shallow } from "enzyme";
import Modal from "../../components/Modal";
import ModalModel from "../../models/ModalModel";

test("should render Modal with loading screen", () => {
  const model = new ModalModel();
  model.visible = true;
  model.loading = true;
  const wrapper = shallow(<Modal modal={model} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render Modal with only text", () => {
  const model = new ModalModel();
  model.visible = true;
  model.text = "Title";
  const wrapper = shallow(<Modal modal={model} />);
  expect(wrapper).toMatchSnapshot();
});

describe("Modal with text and button", () => {
  const model = new ModalModel();
  model.visible = true;
  model.text = "Title";
  model.button = true;
  model.buttonText = "Click";
  const wrapper = shallow(<Modal modal={model} />);

  test("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should hide Modal on button click", () => {
    wrapper.find("button").simulate("click");
    expect(model.visible).toBeFalsy();
  });
});