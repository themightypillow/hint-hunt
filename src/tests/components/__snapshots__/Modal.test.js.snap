// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal with text and button should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modal
    modal={
      ModalModel {
        "button": true,
        "buttonText": "Click",
        "loading": false,
        "text": "Title",
        "visible": true,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "ModalModel@1.keys",
            "observers": Set {},
          },
          "name": "ModalModel@1",
          "target": [Circular],
          "values": Map {
            "visible" => true,
            "text" => "Title",
            "button" => true,
            "buttonText" => "Click",
            "loading" => false,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="modal_text"
      >
        Title
        <button
          onClick={[Function]}
        >
          Click
        </button>
      </div>,
      "className": "modal",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Title",
          <button
            onClick={[Function]}
          >
            Click
          </button>,
        ],
        "className": "modal_text",
      },
      "ref": null,
      "rendered": Array [
        "Title",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Click",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Click",
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="modal_text"
        >
          Title
          <button
            onClick={[Function]}
          >
            Click
          </button>
        </div>,
        "className": "modal",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Title",
            <button
              onClick={[Function]}
            >
              Click
            </button>,
          ],
          "className": "modal_text",
        },
        "ref": null,
        "rendered": Array [
          "Title",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Click",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Click",
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render Modal with loading screen 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modal
    modal={
      ModalModel {
        "button": false,
        "buttonText": "",
        "loading": true,
        "text": "",
        "visible": true,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "ModalModel@3.keys",
            "observers": Set {},
          },
          "name": "ModalModel@3",
          "target": [Circular],
          "values": Map {
            "visible" => true,
            "text" => "",
            "button" => false,
            "buttonText" => "",
            "loading" => true,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="modal_text"
      >
        <Loader />
      </div>,
      "className": "modal",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Loader />,
        "className": "modal_text",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="modal_text"
        >
          <Loader />
        </div>,
        "className": "modal",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Loader />,
          "className": "modal_text",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`should render Modal with only text 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Modal
    modal={
      ModalModel {
        "button": false,
        "buttonText": "",
        "loading": false,
        "text": "Title",
        "visible": true,
        Symbol(mobx did run lazy initializers): true,
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "ModalModel@5.keys",
            "observers": Set {},
          },
          "name": "ModalModel@5",
          "target": [Circular],
          "values": Map {
            "visible" => true,
            "text" => "Title",
            "button" => false,
            "buttonText" => "",
            "loading" => false,
          },
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="modal_text"
      >
        Title
      </div>,
      "className": "modal",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Title",
          false,
        ],
        "className": "modal_text",
      },
      "ref": null,
      "rendered": Array [
        "Title",
        false,
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="modal_text"
        >
          Title
        </div>,
        "className": "modal",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Title",
            false,
          ],
          "className": "modal_text",
        },
        "ref": null,
        "rendered": Array [
          "Title",
          false,
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
